src/main/scala/org/apache/magellan/GeoJSONRelation.scala

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.magellan
+
+import org.apache.spark.sql.catalyst.expressions.GenericMutableRow
+import org.json4s._
+import org.json4s.jackson.JsonMethods._
+
+import org.apache.hadoop.io.{NullWritable, Text}
+import org.apache.magellan.mapreduce.WholeFileInputFormat
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.sources.{BaseRelation, TableScan}
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{Row, SQLContext}
+
+/**
+ * A GeoJSON relation is the entry point for working with GeoJSON formats.
+ */
+case class GeoJSONRelation(path: String)
+                          (@transient val sqlContext: SQLContext)
+  extends BaseRelation with TableScan {
+
+  @transient val sc = sqlContext.sparkContext
+
+  override val schema = {
+    StructType(List(StructField("point", new PointUDT(), true),
+      StructField("polyline", new PolyLineUDT(), true),
+      StructField("polygon", new PolygonUDT(), true),
+      StructField("metadata", MapType(StringType, StringType, true), true),
+      StructField("valid", BooleanType, true)
+    ))
+  }
+
+  override def buildScan(): RDD[Row] = {
+    val numFields = schema.fields.length
+    sc.newAPIHadoopFile(
+      path,
+      classOf[WholeFileInputFormat],
+      classOf[NullWritable],
+      classOf[Text]
+    ).flatMap { case(k, v) =>
+      val line = v.toString()
+      parseShapeWithMeta(line)
+    }.mapPartitions { iter =>
+      val row = new GenericMutableRow(numFields)
+      iter.flatMap { case (shape: Shape, meta: Option[Map[String, String]]) =>
+        (0 until numFields).foreach(i => row.setNullAt(i))
+        row(3) = meta.fold(Map[String, String]())(identity)
+        row(4) = if (shape == NullShape) false else shape.isValid()
+        shape match {
+          case NullShape => None
+          case _: Point =>
+            row(0) = shape
+            Some(row)
+          case _: PolyLine =>
+            row(1) = shape
+            Some(row)
+          case _: Polygon =>
+            row(2) = shape
+            Some(row)
+          case _ => ???
```

Comment:
the content of this flat map looks identical to the one in [ShapefileRelation](https://github.com/harsha2010/magellan/blob/ff45b0bd0bff21ddaa345343f886c2ed9cf78760/src/main/scala/org/apache/magellan/ShapefileRelation.scala#L80). Could we abstract these somehow?